src/pymoca/ast.py

```diff
@@ -908,45 +926,111 @@ def __deepcopy__(self, memo):
+    ):
+        """ast_ref is a reference to the AST node where this instance is defined.
+        All named keyword arguments optional for backward compatibility."""
+
+        # super().__init__() is only needed if 1st in method resolution order
+        super().__init__(**kwargs)
+
+        self.ast_ref = ast_ref
+
+        if modification_environment is not None:
+            self.modification_environment = modification_environment
+        else:
+            self.modification_environment = ClassModification()
+
+        if "name" in kwargs:
+            self.name = kwargs["name"]
+        elif ast_ref is not None:
+            self.name = ast_ref.name
+        else:
+            self.name = ""  # The default in Symbol
+
+        if "type" in kwargs:
+            self.type = kwargs["type"]
+        elif ast_ref is not None:
+            self.type = ast_ref.type
+        else:
+            self.type = ComponentRef()  # The default in Symbol
+
+        self.fully_instantiated = fully_instantiated
+
+    def __repr__(self):
+        return f"name={self.name!r}, ast_ref={self.ast_ref!r}, modification_environment={self.modification_environment!r}"
+
+
+class InstanceClass(InstanceElement, Class):
     """
-    Class used during instantiation/expansion of the model. Modififcations on
+    Class used during instantiation/expansion of the model. Modifications on
     symbols and extends clauses are shifted to the modification environment of
     this InstanceClass.
     """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.modification_environment = ClassModification()
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
 
     def __repr__(self):
-        return "{}(type={!r}, name={!r}, modification_environment={!r})".format(
-            type(self).__name__, self.type, self.name, self.modification_environment
-        )
+        return f"{type(self).__name__}({super().__repr__()!s})"
 
 
-class Tree(Class):
+class InstanceSymbol(InstanceElement, Symbol):
     """
-    The root class.
+    Symbol used during instantiation/expansion of the model.
     """
 
-    def extend(self, other: "Tree") -> None:
-        self._extend(other)
-        self.update_parent_refs()
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
 
-    def _update_parent_refs(self, parent: Class) -> None:
-        for c in parent.classes.values():
-            c.parent = parent
-            self._update_parent_refs(c)
+    def __repr__(self):
+        return f"{type(self).__name__}({super().__repr__()!s})"
 
-    def update_parent_refs(self) -> None:
-        self._update_parent_refs(self)
+
+class InstanceExtends(InstanceElement, Class):
+    """
+    Placeholder for extends during instantiation/expansion of the model.
+    This is the "unnamed node" referenced in the language spec.
+    """
+
+    def __init__(self, visibility: Optional[Visibility] = Visibility.PUBLIC, **kwargs):
```

Comment:
Why does an extends clause have visibility?

Reply:
Because it can be included in a `protected` section of a class definition which would limit the visibility of everything in the extends clause to only the containing class and classes that inherit from the containing class. In simulation tools, `protected` elements are by default hidden from the user.

Our `ast.Class` should also have a visibility attribute because it can defined in a `protected` section. I have a TODO in there related to this.